[CDAP 4338] Do not expire datasets based on GC's weak references; introduce a new API dismissDataset()

anew opened the pull request:
Please read https://issues.cask.co/browse/CDAP-4338 before review (note: branch name incorrectly has 4388). 
- removes the weak values configuration from the dataset cache, along with all the code to support that
- adds a new API DatasetProvider. It will replace DatasetContext in the future and has additional methods to explicitly release/dismiss a dataset that is no longer used
- all program contexts now implement DatasetProvider
- modified one example (FileSetExample) that uses dynamic datasets, to dismiss its service's datasets after use, and extended the test case to make repeated calls to the service.
- still pending: documentation changes to document the new behavior and API method. I will do that in its own commit so that it can be reviewed on its own

Docs Build: http://builds.cask.co/browse/CDAP-DOB66-5 [Updated for Build 5].

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/data/DatasetProvider.java:
```diff
@@ -0,0 +1,51 @@
+
+import co.cask.cdap.api.dataset.Dataset;
+
+import java.util.Map;
```
unused import?

chtyim replied:
Oh, it's used in the javadoc. Nevermind

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/data/DatasetProvider.java:
```diff
@@ -0,0 +1,51 @@
+   * or {@link #getDataset(String, Map)}. It is up to the implementation of the
+   * provider whether this dataset is discarded, or whether it is reused as the
+   * result for subsequent {@link #getDataset} calls with the same arguments.
+
```
Missing `*`?

anew replied:
oops

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/data/DatasetProvider.java:
```diff
@@ -0,0 +1,51 @@
+   *
+   * @param dataset The dataset to be dismissed.
+   */
+  void dismissDataset(Dataset dataset);
```
Maybe better call `discardDataset`?

anew replied:
It is subtle but I used "dismiss" to mean "mark to be closed and discarded". I am open to renaming. @awholegunch what does the English teacher say?

awholegunch replied:
Your Javadoc says "the DatasetProvider must close and discard it." That would suggest that "discardDataset" is a better choice. When someone is dismissed, they are freed from their current obligations, similar to "released".

But it sounds like you want to force the recovery of the resources, so making that clearer by using "discardDataset" would reinforce that.

jwang47 replied:
Just by reading the javadocs on this interface, I'm not sure if I understand the difference between `releaseDataset` and `dismissDataset`. Is the only difference that

`releaseDataset` allows the `DatasetProvider` to close or reuse, and
`dismissDataset` allows the `DatasetProvider` to close (no reuse)?

anew replied:
That is correct, Alvin, The idea is that in the future we want to be able to have time-based expiration. A program can "release" the dataset, allowing the system to discard it at its discretion. Whereas "dismiss" means that it will be discard as soon as possible, to free resources right away. In a way, "release" is like Java's GC, with no guarantee as to how soon it will be discarded; whereas "dismiss" is like C++'s delete(), forcing immediate freeing of the resources.

jwang47 replied:
Then maybe it makes sense to have a single method:

```
releaseDataset(Dataset dataset, boolean allowReuse);
```

anew replied:
@awholegunch I can easily rename it to discard. Just want to make sure that you understand: calling dismiss() does not immediately discard it, it only marks it for being discarded. That can only happen after the current transaction completes, otherwise that dataset's commitTx(0 would be called after the dataset is closed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java:
```diff
@@ -254,6 +254,18 @@ private void initializeTransactionAwares() {
+  }
+
+
+
```
Seems like too many extra lines?

anew replied:
Weird, I don't see them in my IntelliJ.

anew replied:
Oh, now I do. Fixed.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java:
```diff
@@ -129,6 +124,11 @@ public DynamicDatasetCache(SystemDatasetInstantiator instantiator,
 
+  @Override
+  public void releaseDataset(Dataset dataset) {
+    dismissDataset(dataset);
```
So the current implementation is not to reuse a dataset instance once it is released?

anew replied:
yes. Since the javadoc states: it is up to the implementation, I started with the simplest possible one. Another option would be to remove releaseDataset() for now. Only that that will break compatibility again when we introduce it.

chtyim replied:
That's no problem. Just to clarify :)

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -200,21 +184,53 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
+      }
+    }
+    // we can only hope that dataset.toString() is meaningful
+    LOG.warn("Attempt to dismiss dataset that was not acquired through this provider: " + dataset);
```
Use `{}` instead of `+`.

anew replied:
Ugh, yeah, this was an exception which does not support {}. Will fix

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -200,21 +184,53 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
+      return;
+    }
+    if (txContext == null || !(dataset instanceof TransactionAware)) {
+      dismissSafely(dataset);
```
So it seems like only `dismissSafely` will invalid the dataset from the cache, but not in `dismissAfterTx`. Do it miss some of the logic?

anew replied:
Yes rest of the logic is in the transaction context (DelayedDismissingTxContext). We need to hold on to the dataset until the transaction completes. Only the tx context knows when that happens. It overrides abort() and finish() to call dismissSafely() for all of the datasets that were marked for that using dismissAfterTx()

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -169,30 +166,20 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
-                            dataset, key, actual));
-          }
+        // this better be the same dataset, otherwise the cache did not work
+        if (existing != dataset) {
```
The else and if scopes can be combine to one `else-if` to make the code less clutter.

anew replied:
Yes, that was a relic from old code structure. Fixed.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -200,21 +187,54 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
+    }
+    // remove from activeTxAwares in any case - a dismissed dataset does not need to participate in external tx
+    // iterates over all datasets but we do not expect this map to become large
+    for (Map.Entry<DatasetCacheKey, TransactionAware> entry : activeTxAwares.entrySet()) {
```
Maybe cleaner by using `activeTxAwares.values().remove(dataset)`.

chtyim replied:
hum, `Collections.remove` uses `equals`. So, nevermind. (or you have to use `Iterators.removeIf` with a predicate that use `==` check.

anew replied:
Also, if I try that, IntelliJ warns me about a suspicious method call... Since the complexity is the same (requires iterating over the map), I find it ok to leave the explicit for loop here.

gokulavasan commented on cdap-docs/developers-manual/source/building-blocks/datasets/index.rst:
```diff
@@ -125,8 +108,29 @@ There are two ways to use a dataset in a program:
+  However, contrary to static datasets, dynamic datasets allow the release of the resources held by their Java classes
+  after you are done using them. You can do that by calling the ``dismissDataset()`` method of the program context:
+  it marks the dataset to be closed and removed from all transactions as soon as the current transaction completes,
+  thereby releasing its resources.
```
Is there a performance (or other) reason why the user should use ``dismissDataset()`` whenever possible ?

anew replied:
Yes, as outlined here, to release resources. 
Not sure whether I misread the question....

gokulavasan replied:
I was wondering if we should mention that explicitly in the docs.

anew replied:
It is explicit: "However, contrary to static datasets, dynamic datasets allow the release of the resources held by their Java classes"

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java:
```diff
@@ -200,21 +187,54 @@ private void closeDataset(DatasetCacheKey key, Dataset dataset) {
-    for (DatasetCacheKey key : expiredTxAwares) {
-      activeTxAwares.remove(key);
+    for (Map.Entry<DatasetCacheKey, TransactionAware> entry : activeTxAwares.entrySet()) {
+      txContext.addTransactionAware(entry.getValue());
```
Seems like we only need the value. Maybe better just to iterator over `activeTxAwares.values()`. Or maybe have a constructor of `DelayedDismissingTransactionContext` that takes a `Iterables<TransactionAware>`?

anew replied:
Yeah that ctor is now possible. The old code had to do extra work around this... now we can simplify.

anew replied:
done

gokulavasan commented on cdap-docs/developers-manual/source/building-blocks/datasets/index.rst:
```diff
@@ -125,8 +108,29 @@ There are two ways to use a dataset in a program:
+  thereby releasing its resources.
+
+  If you know the dataset name at the time the program starts, you can store a reference to the dataset in a member
+  variable at that time (similar to static datasets, but assigned explicitly by you)::
```
Just to confirm, these datasets are treated as static datasets and so the instance will not expire?

anew replied:
The instance will expire if you call dismissDataset()

chtyim commented:
Took a pass of the code. I believe the logic is correct (it's rather complicated). Only couple minor comments about some code simplification.

awholegunch commented:
I've made two commits; please check. LGTM, aside from :
- the naming issue above (*dismissDataset* versus *discardDataset*)
- have we made it clear enough when it is better to use release versus dismissDataset/discardDataset?

I can see that if my app is done with a dataset, I can call ``discardDataset``, and its resources will be released when the tx completes. But when would I get any advantage to using ``release``? We don't show an example or use case of when that might be useful.

anew commented:
The releaseDataset() is currently not documented in the developer guid, only in the Javadocs. We currently always dismiss (release() calls dismiss()). However, we wanted to create that API because every time we add a method to an interface, we break backward compatibility. This time we are deprecating DatasetContext and adding DatasetProvider (which extends the former). I was trying to add both methods to the API even if, for now, their implementation is the same. Otherwise we will have to break the compatibility again next time. 

The use case is simple: If I have a generic service that needs to determine the dataset name for every request, then it can't use static datasets. It also may not want to keep all datasets around until it terminates, because it may run out of resources. So it will call dismissDataset(). But if I repeatedly call that service for the same dataset, then it has to create a new instance every time, which is expensive and slow. Therefore we want to have a way to say: System, you can discard this when you think it is the right time to do so, I am done with it. Now the system may implement a time-based expiry or LRU, or something similar, and we can add configuration options for that later. If the program gets the same dataset again before it has actually been discarded, then it can be reused. Otherwise, the system can instantiate a new one. 

Whereas, in a similar scenario, if the user knows that the dataset uses resources extensively (perhaps resources of an external system), and he knows that the dataset will only be used occasionally, then he can call dismiss to explicitly force discarding and freeing the resources as soon as possible.

anew commented:
Regarding the naming issue, I have no concerns with renaming. Will do that now.

anew commented:
@awholegunch you changes to the docs looks good to me.

Regarding releaseDataset(), I really don't want to add that prose to the docs. We have two choices:
1. remove this interface and bite the bullet and break the api again when we introduce it
2. document in its javadoc that the current implementation is the same as discardDataset()

Opinions?

anew commented:
OK, I have added some clarification around releaseDataset() vs. discardDataset(). @awholegunch, can you please take a look at the last commit? 

Other than that, I believe I have addressed all comments. Is this ready to merge then?

anew commented:
By the way, the build for the latest state of this PR is running here: http://builds.cask.co/browse/CDAP-DUT3290-6

awholegunch commented:
I have started a new doc build to check, and have changed the usage of "file set" to "FileSet" in two places in the last commit. Please check. Otherwise, assuming it passes build 5, LGTM.

awholegunch commented:
Doc build is broken in develop (see https://github.com/caskdata/cdap/pull/4705) and this doc build is broken with the same errors. It is probably fine. However, the Javadocs have not been run; this can be fixed in the release branch if necessary. I have added a note the Issue about checking the Javadocs.

anew commented:
The last doc build I ran on my branch passed. The latest failure must be because some change went into develop that breaks it. I don't know what to do about that. Develop must be fixed.

anew commented:
@awholegunch your changes are ok to me. 
@gokulavasan @jwang47 please take a look and indicate whether you think that the documentation is now clear enough. 
I will merge as soon as you give me thumbs-up.

awholegunch commented:
@anew I have opened a PR ( https://github.com/caskdata/cdap/pull/4705 ) to fix the doc build in develop.

gokulavasan commented:
Looks GTM.

anew commented:
OK, thanks. I will merge this unless @jwang47 would like another LGTM for the code changes (Terence reviewed it but had to leave before I addressed all comments).

anwar6953 commented:
I think it's GTM!

anew commented:
ok, merging.

anwar6953 commented:
deprecated/beta annotation markings LGTM
